pkg/provisioner/ironic/ironic.go

```diff
@@ -1227,3 +1260,41 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// softPowerOff sends 'soft power off' request to BM node.
+// If soft power off is not supported, the request ends with an error.
+// Otherwise the request ends with no error and the result should be
+// checked later via node fields "power_state", "target_power_state"
+// and "last_error".
+func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error) {
+	p.log.Info("ensuring host is powered off by \"soft power off\" command")
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+
+	if ironicNode.PowerState != powerOff {
+		targetState := ironicNode.TargetPowerState
+		// If the target state is either powerOff or softPowerOff, then we should wait
+		if targetState == powerOff || targetState == softPowerOff {
+			p.log.Info("waiting for power status to change")
+			result.RequeueAfter = powerRequeueDelay
+			result.Dirty = true
+			return result, nil
+		}
+		// If the target state is unset while the last error is set,
+		// then the last execution of soft power off has failed.
+		if targetState == "" && ironicNode.LastError != "" {
+			return result, SoftPowerOffFailed{Address: p.host.Spec.BMC.Address}
```

Comment:
This just logs an error and requeues. There's no reason to think we won't just end up back here in an infinite loop, is there?
This is where we want to do a hard power off I think.

Reply:
I do n’t know if you marked the wrong code. In the code you marked, `SoftPowerOffFailed` is returned, and then go back to execute `hardPowerOff`.
```
		// In case of soft power off is unsupported or has failed,
		// we activate hard power off.
		case SoftPowerOffUnsupportedError, SoftPowerOffFailed:
			return p.hardPowerOff()
```

Reply:
Sorry, I got confused about what gets called from where. You are correct.